lib/Monads/trace/Trace_More_RG.thy

```diff
@@ -0,0 +1,685 @@
+  assumes "\<And>x xs. \<lbrace>P_B\<rbrace>,\<lbrace>R\<rbrace> b x xs \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>"
+  shows
+  "\<lbrace>P_A and P_B\<rbrace>,\<lbrace>R\<rbrace>
+   case ts of
+       [] \<Rightarrow> a
+     | x#xs \<Rightarrow> b x xs
+   \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>"
+  apply (cases ts)
+  apply (simp, rule rg_weaken_pre, rule assms, simp)+
+  done
+
+lemma rg_vcg_if_lift2:
+  "\<lbrace>R\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. (P rv s0 s \<longrightarrow> X rv s0 s) \<and> (\<not> P rv s0 s \<longrightarrow> Y rv s0 s)\<rbrace> \<Longrightarrow>
+   \<lbrace>R\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. if P rv s0 s then X rv s0 s else Y rv s0 s\<rbrace>"
+
+  "\<lbrace>R\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. (P' rv \<longrightarrow> X rv s0 s) \<and> (\<not> P' rv \<longrightarrow> Y rv s0 s)\<rbrace> \<Longrightarrow>
+   \<lbrace>R\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv. if P' rv then X rv else Y rv\<rbrace>"
+  by (auto simp: validI_def)
+
+lemma rg_vcg_if_lift_ER: (* Required because of lack of rv in lifting rules *)
+  "\<lbrace>R\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. (P rv s0 s \<longrightarrow> X rv s0 s) \<and> (\<not> P rv s0 s \<longrightarrow> Y rv s0 s)\<rbrace>,\<lbrace>E\<rbrace> \<Longrightarrow>
+   \<lbrace>R\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. if P rv s0 s then X rv s0 s else Y rv s0 s\<rbrace>,\<lbrace>E\<rbrace>"
+
+  "\<lbrace>R\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. (P' rv \<longrightarrow> X rv s0 s) \<and> (\<not> P' rv \<longrightarrow> Y rv s0 s)\<rbrace>,\<lbrace>E\<rbrace> \<Longrightarrow>
+   \<lbrace>R\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv. if P' rv then X rv else Y rv\<rbrace>,\<lbrace>E\<rbrace>"
+  by (auto simp: validI_def validIE_def)
+
+lemma rg_list_all_lift:
+  "\<lbrakk>\<And>r. r \<in> set xs \<Longrightarrow> \<lbrace>Q r\<rbrace>,\<lbrace>R\<rbrace> f -,\<lbrace>\<lambda>rv. Q r\<rbrace>; \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<top>\<top>\<top>\<rbrace>\<rbrakk>
+   \<Longrightarrow> \<lbrace>\<lambda>s0 s. list_all (\<lambda>r. Q r s0 s) xs \<and> P s0 s\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. list_all (\<lambda>r. Q r s0 s) xs\<rbrace>"
+  apply (rule validI_split[rotated, simplified pred_conj_def])
+   apply assumption
+  apply (induct xs; simp)
+   apply wpsimp
+  apply (rule rg_vcg_conj_lift; simp)
+  done
+
+\<comment> \<open>FIXME MC: no longer true since it requires "prefix_closed undefined", is that a problem?
+lemma undefined_valid: "\<lbrace>\<bottom>\<bottom>\<rbrace>,\<lbrace>R\<rbrace> undefined \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>"
+  by (rule rg_pre_cont)\<close>
```

Comment:
Interesting. It used to be convenient to get `False` for things like `haskell_error` (which maps to `undefined`) so we had to prove that the case is unreachable.

Do we have a better concept that we could map `haskell_error` to? I guess we could map `error` to produce chaos + failure, i.e. the monad that does no trace step, but produce UNIV as possible results and next states. That would actually be better on the nondet monad as well, so that we don't get these random `undefined` instances.

Reply:
(so, yes, delete -- but we need to leave a note somewhere what to do with `haskell_error`)